Switching to the default connection class to address slower connection with RequestsHttpConnection

beaioun opened the pull request:
### Description
This PR implements the default Urllib3HttpConnection to replace RequestsHttpConnection to address the slower connection issue, as recommended by the opensearch-py guide for connection classes. Upon checking the code, I discovered this was the only place RequestsHttpConnection was used. Please do point out if I was wrong here.

### Issues Resolved
#408 

### Testing
- [ ] New functionality includes testing


---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

IanHoang reviewed (changes requested):
With the new change, it's failing for a unittest that tests client creation with aws credentials. Seems like it's failing in the underlying packages. We'll need to see why this only happens when Urllib3HttpConnection is used and not when RequestsHttpConnection is used.
```
s = <opensearchpy.helpers.signer.AWSV4SignerAuth object at 0x10b73c940>

    def b(s):
>       return s.encode("latin-1")
E       AttributeError: 'AWSV4SignerAuth' object has no attribute 'encode'

.venv/lib/python3.8/site-packages/urllib3/packages/six.py:687: AttributeError
```

beaioun commented:
Hey @IanHoang , I figured out where the issue was. The Urllib3HttpConnection has to pair with Urllib3AWSV4SignerAuth for http_auth because that is callable, whereas RequestsAWSV4SignerAuth is not and will try to call make_headers in the urllib3 and throws an error for the subsequent actions. Could you try to run the workflow again? Thanks.

rishabh6788 commented:
Also, I don't think `opensearch-py-2.3.2` has support for the new methods you added. 
Can you please update the `opensearch-py` version to 2.4.1 , test it locally and update the PR?
@beaioun

beaioun commented:
> Also, I don't think `opensearch-py-2.3.2` has support for the new methods you added.
> Can you please update the `opensearch-py` version to 2.4.1 , test it locally and update the PR?

@rishabh6788 I see, I'll update to 2.4.1 see if that works.

rishabh6788 commented:
FYI, there is bug in `opensearch-py-2.4.1` where it is not working with OpenSearch Serverless. 
https://github.com/opensearch-project/opensearch-py/issues/600

dblock commented on osbenchmark/client.py:
```diff
@@ -291,10 +291,10 @@ def create(self):
                                                 self.aws_log_in_dict["service"])
         return opensearchpy.OpenSearch(hosts=self.hosts, use_ssl=True, verify_certs=True, http_auth=aws_auth,
-                                       connection_class=opensearchpy.RequestsHttpConnection)
+                                       connection_class=opensearchpy.Urllib3HttpConnection)
```
FYI this is now the default and you could remove this option if you want to. Doesn't hurt to keep it either.

rishabh6788 replied:
Thank you @dblock , so if we just use `opensearch-py-2.4.1` then we don't need to explicitly specify `connection_class` parameter?

dblock replied:
Yes

dblock commented:
> FYI, there is bug in `opensearch-py-2.4.1` where it is not working with OpenSearch Serverless.
> [opensearch-project/opensearch-py#600](https://github.com/opensearch-project/opensearch-py/issues/600)

This is not the case, it works with AOSS well. The LangChain integration was failing because it [checks for `signer.service` explicitly and that field was removed](https://github.com/langchain-ai/langchain/blob/b4312aac5c0567088353178fb70fdb356b372e12/libs/langchain/langchain/vectorstores/opensearch_vector_search.py#L83).

beaioun commented:
> Also, I don't think `opensearch-py-2.3.2` has support for the new methods you added. Can you please update the `opensearch-py` version to 2.4.1 , test it locally and update the PR? @beaioun

@rishabh6788 I updated the opensearch-py version to 2.4.1 and fixed some other importing issues with this version. All tests passed on my end. 769ff90bc024efbe725b4d11c14b0781a137d63c
